plaso/parsers/setupapi.py

```diff
@@ -63,74 +65,109 @@ class SetupapiLogParser(text_parser.PyparsingMultiLineTextParser):
-      pyparsing.SkipTo(pyparsing.lineEnd()) +
-      pyparsing.ZeroOrMore(pyparsing.lineEnd()))
+      pyparsing.lineEnd())
+
+  # See https://docs.microsoft.com/en-us/windows-hardware/drivers/install/format-of-a-text-log-section-footer
+  _SECTION_END_EXIT_STATUS = (
+      pyparsing.Literal('<<<  [Exit status: ').suppress() +
+      pyparsing.CharsNotIn(']').setResultsName('exit_status') +
+      pyparsing.Literal(']') +
+      pyparsing.lineEnd())
+
+  # See https://docs.microsoft.com/en-us/windows-hardware/drivers/install/format-of-log-entries-that-are-not-part-of-a-text-log-section
+  _SECTION_BODY_LINE = (
+      pyparsing.stringStart +
+      pyparsing.MatchFirst([
+          pyparsing.Literal('!!!  '),
+          pyparsing.Literal('!    '),
+          pyparsing.Literal('     ')]) +
+      pyparsing.restOfLine).leaveWhitespace()
+
+  # See https://docs.microsoft.com/en-us/windows-hardware/drivers/install/format-of-log-entries-that-are-not-part-of-a-text-log-section
+  _NON_SECTION_LINE = (
+      pyparsing.stringStart +
+      pyparsing.MatchFirst([
+          pyparsing.Literal('   . '),
+          pyparsing.Literal('!!!  '),
+          pyparsing.Literal('!    '),
+          pyparsing.Literal('     ')]) +
+      pyparsing.restOfLine).leaveWhitespace()
+
+  # These lines do not appear to be documented in the Microsoft documentation.
+  _BOOT_SESSION_LINE = (
+      pyparsing.Literal('[Boot Session:') +
+      _SETUPAPI_DATE_TIME +
+      pyparsing.Literal(']'))
+
+  # pylint: enable=line-too-long
 
   LINE_STRUCTURES = [
-      ('logline', _SETUPAPI_LINE),
-  ]
+      ('ignorable_line', _BOOT_SESSION_LINE),
+      ('ignorable_line', _LOG_HEADER_END),
+      ('ignorable_line', _LOG_HEADER_START),
+      ('ignorable_line', _NON_SECTION_LINE),
+      ('ignorable_line', _SECTION_BODY_LINE),
+      ('section_end', _SECTION_END),
+      ('section_end_exit_status', _SECTION_END_EXIT_STATUS),
+      ('section_header', _SECTION_HEADER),
+      ('section_start', _SECTION_HEADER_START)]
+
+  def __init__(self):
+    """Initializes a setupapi parser."""
+    super(SetupapiLogParser, self).__init__()
+    self._last_end_time = None
+    self._last_entry_type = None
 
-  def _ParseRecordLogline(self, parser_mediator, structure):
-    """Parses a logline record structure and produces events.
+  def _BuildDatetime(self, time_structure):
+    """Builds time elements from a setupapi time stamp.
 
     Args:
-      parser_mediator (ParserMediator): mediates interactions between parsers
-          and other components, such as storage and dfvfs.
-      structure (pyparsing.ParseResults): structure of tokens derived from
-          log entry.
+      time_structure (pyparsing.ParseResults): structure of tokens derived from
+          a setupapi time stamp.
+
+    Returns:
+      dfdatetime.TimeElements: date and time extracted from the value or None
+          if the value does not represent a valid string.
```

Comment:
a valid string => a valid date and time value

Reply:
Done.